File: codebase/core/main/src/Primitives/TreeNode.fs
namespace Forest

open Forest.NullHandling

open System
open System.Text
open System.Diagnostics
open System.Runtime.CompilerServices


[<DebuggerDisplay("{this.ToString()}")>]
[<CustomComparison>]
[<CustomEquality>]
type TreeNode = 
    | [<DebuggerBrowsable(DebuggerBrowsableState.Never)>] Shell_
    | [<DebuggerBrowsable(DebuggerBrowsableState.Never)>] ViewID_ of parent:TreeNode * region:rname * view:vname * hash:hash
    [<CompiledName("Shell")>]
    static member shell = 
        TreeNode.Shell_
    [<CompiledName("NewKey")>]
    static member newKey region view parent = 
        TreeNode.ViewID_(parent, region, view, Fuid.newID().Hash)
    member this.Parent 
        with get() = match this with Shell_ -> Shell_ | ViewID_ (p, _, _, _) -> p
    member this.Region 
        with get() = match this with Shell_ -> rname.Empty | ViewID_ (_, r, _, _) -> r
    member this.View 
        with get() = match this with Shell_ -> vname.Empty | ViewID_ (_, _, v, _) -> v
    member this.Hash 
        with get() = match this with Shell_ -> Fuid.empty.Hash | ViewID_ (_, _, _, h) -> h
    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    member private this.cmp (other:TreeNode) = 
        StringComparer.Ordinal.Compare(this.Hash, other.Hash)
    [<MethodImpl(MethodImplOptions.AggressiveInlining)>] 
    member private this.eq (other: TreeNode) = 
        StringComparer.Ordinal.Equals(this.Hash, other.Hash)
    override this.Equals o = 
        match null2opt o with
        | Some v ->
            match v with
            | :? TreeNode as other -> this.eq other
            | _ -> false
        | None -> false
    override this.GetHashCode() = 
        this.Hash.GetHashCode()
    member private this.ToStringBuilder(stopAtRegion:bool):StringBuilder =
        let sb =
            match this with 
            | ViewID_ (p, r, v, h) -> 
                let sb:StringBuilder = p.ToStringBuilder(false).Append((if r.Length = 0 then "shell" else r))
                if (stopAtRegion) then sb
                else sb.AppendFormat("/{0}({1})", h, v)
            | Shell_ -> StringBuilder("")
        sb.Append('/')
    member this.RegionFragment 
        with get() = this.ToStringBuilder(true).ToString()
    override this.ToString() = 
        this.ToStringBuilder(false).ToString()
    interface IComparable<TreeNode> with 
        member this.CompareTo other = this.cmp other
    interface IComparable with
        member this.CompareTo o = 
            match o with
            | :? TreeNode as id -> this.cmp id
            | :? IComparable as c -> (-1)*(c.CompareTo this)
            | _ -> raise (NotSupportedException ())
    interface IEquatable<TreeNode> with 
        member this.Equals other = this.eq other

module internal TreeNode =
    //let add id region name parent = HierarchyKey.ViewID_(parent, region, name, id)
    //let addNew region name parent = add (Fuid.newID().Hash) region name parent
    let isShell (id:TreeNode) = match id with TreeNode.Shell_ -> true | _ -> false
    

File: codebase/core/main/src/Tree.Node.cs
using System;
using System.Diagnostics;
using System.Runtime.InteropServices;
using Forest.ComponentModel;

namespace Forest
{
    partial class Tree
    {
        /// <summary>
        /// A struct representing a single node inside a Forest <see cref="Tree"/>.
        /// </summary>
        #if NETSTANDARD2_0_OR_NEWER || NETFRAMEWORK
        [Serializable]
        #endif
        [StructLayout(LayoutKind.Sequential)]
        [DebuggerDisplay("{this." + nameof(ToString) + "()}")]
        public struct Node : IComparable<Node>, IEquatable<Node>
        {
            internal static Node Create(string key, ViewHandle viewHandle, string region, object model, Node parent)
            {
                var viewState = model == null ? null : new ViewState?(Forest.ViewState.Create(model));
                
                return new Node(parent.Key, region, viewHandle, key, viewState, 0u);
            }

            public static readonly Node Shell;
            
            private readonly string _region;
            private readonly ViewHandle _viewHandle;
            private readonly string _key;
            private readonly string _parentKey;

            internal Node(Node parent, string region, ViewHandle viewHandle, string key) 
                : this(parent.Key, region, viewHandle, key, null, 0u) { }
            private Node(string parentKey, string region, ViewHandle viewHandle, string key, ViewState? viewState, uint revision)
            {
                _key = key;
                _parentKey = parentKey;
                _region = region;
                _viewHandle = viewHandle;
                ViewState = viewState;
                Revision = revision;
            }
            
            int IComparable<Node>.CompareTo(Node other)
            {
                return StringComparer.Ordinal.Compare(Key, other.Key);
            }

            /// <inheritdoc />
            public bool Equals(Node other)
            {
                return StringComparer.Ordinal.Equals(Key, other.Key);
            }

            /// <inheritdoc />
            public override bool Equals(object obj) => obj is Node other && Equals(other);

            /// <inheritdoc />
            public override int GetHashCode() => (Key != null ? Key.GetHashCode() : 0);

            internal Node UpdateRevision() => new Node(ParentKey, Region, ViewHandle, Key, ViewState, 1u + Revision);

            internal Node SetViewState(ViewState viewState)
            {
                return new Node(ParentKey, Region, ViewHandle, Key, viewState, 1u + Revision);
            }

            // private StringBuilder ToStringBuilder(bool stopAtRegion)
            // {
            //     var sb = Parent == null
            //         ? new StringBuilder() 
            //         : Parent.ToStringBuilder(false).Append(Region.Length == 0 ? "shell" : Region);
            //     return (stopAtRegion ? sb : sb.AppendFormat("/{0} #{1}", ViewHandle, Key)).Append('/');
            // }

            //public override string ToString() => ToStringBuilder(false).ToString();

            public string ParentKey => _parentKey ?? string.Empty;

            /// <summary>
            /// Gets the name of the <see cref="IRegion"/> that contains the view represented by this <see cref="Node"/>.
            /// </summary>
            public string Region => _region ?? string.Empty;

            /// <summary>
            /// Gets a <see cref="ViewHandle"/> that can be used to obtain the <see cref="IViewDescriptor">view descriptor</see>
            /// </summary>
            public ViewHandle ViewHandle => _viewHandle ?? ViewHandle.FromName(string.Empty);

            /// <summary>
            /// A unique string identifying the current node.
            /// </summary>
            public string Key => _key ?? Guid.Empty.ToString();
            
            public ViewState? ViewState { get; }
            public uint Revision { get; }

            //public string RegionSegment => ToStringBuilder(true).ToString();
        }
    }
}